### PR TITLE
Emit a warning when Miri is used with optimizations

### PR DESCRIPTION
**Pull request by saethlin**
This may address https://github.com/rust-lang/miri/issues/2797, by clarifying to the user what is going on and what the consequences of their choices are.

Fixes https://github.com/rust-lang/miri/issues/2797

### CODE REVIEW

**Comment by oli-obk** [comment]
Should we also tell the user that this will not significantly  improve the performance of running miri except in microbenchmarks?

File: ci.sh [hunk 1]
```diff
@@ -43,7 +43,7 @@ function run_tests {
     # optimizations up all the way, too).
     # Optimizations change diagnostics (mostly backtraces), so we don't check
     # them. Also error locations change so we don't run the failing tests.
-    MIRIFLAGS="${MIRIFLAGS:-} -O -Zmir-opt-level=4" MIRI_SKIP_UI_CHECKS=1 ./miri test -- tests/{pass,panic}
+    MIRIFLAGS="${MIRIFLAGS:-} -O -Zmir-opt-level=4 -Cdebug-assertions=yes" MIRI_SKIP_UI_CHECKS=1 ./miri test -- tests/{pass,panic}
```
Comment (RalfJung) [inline ci.sh:46]: Why add the flag back here? I guess some tests rely on it? Please add a comment.

File: src/bin/miri.rs [hunk 2]
```diff
@@ -82,6 +84,15 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
...
             }
 
+            if tcx.sess.opts.optimize != OptLevel::No {
+                tcx.sess.warn("Miri does not support optimizations.");
+            } else if tcx.sess.mir_opt_level() > 0 {
+                tcx.sess.warn("You have explicitly enabled MIR optimizations, overriding Miri's default \
+                    which is to completely disable them. Any optimizations may hide UB that Miri would \
+                    otherwise detect, and it is not necessarily possible to predict what kind of UB will \
+                    be missed.");
+            }
```
Comment (RalfJung) [inline src/bin/miri.rs:94]: So if the user says `-O -Zmir-opt-level=3`, they will get the first warning, not the second? That seems wrong?

File: ci.sh (hunk 1 above)
Comment (saethlin) [reply ci.sh:48]: Yeah. I considered doing something more clever, such `cfg(not(debug_assertions))` on the test cases, but the tests pretty explicitly call `debug_assert!` so it seemed very circular.

File: src/bin/miri.rs (hunk 2 above)
Comment (saethlin) [reply src/bin/miri.rs:100]: That was definitely wrong.

File: src/bin/miri.rs [hunk 3]
```diff
@@ -82,6 +84,21 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
...
+                    such as whether debug assertions and overflow checks are enabled, those settings are \
+                    still applied.");
+            }
+            if tcx.sess.mir_opt_level() > 0 {
+                tcx.sess.warn("You have explicitly enabled MIR optimizations, overriding Miri's default \
+                    which is to completely disable them. Any optimizations may hide UB that Miri would \
+                    otherwise detect, and it is not necessarily possible to predict what kind of UB will \
+                    be missed. If you are enabling optimizations to make Miri run faster, we advise using \
+                    cfg(miri) to shrink your workload instead. The impact of enabling MIR optimizations is \
+                    usually marginal at best.");
```
Comment (RalfJung) [inline src/bin/miri.rs:99]:
```suggestion
                    cfg(miri) to shrink your workload instead. The performance benefit of enabling MIR \
                    optimizations is usually marginal at best.");
```

**Pull request merged**